asset/asset.go

```diff
@@ -943,6 +944,447 @@ type GroupKeyReveal interface {
+		root:                        rootHash,
+		customSubtreeRoot:           customRoot,
+		customSubtreeInclusionProof: customSubtreeInclusionProof,
+	})
+}
+
+// Validate checks that the group key reveal tapscript is well-formed and
+// compliant.
+func (g *GroupKeyRevealTapscript) Validate(assetID ID) error {
+	// Compute the final tapscript root hash from the genesis asset ID and
+	// the custom tapscript subtree root hash.
+	tapscript, err := NewGroupKeyTapscriptRoot(
+		assetID, g.customSubtreeRoot,
+	).Unpack()
+	if err != nil {
+		return fmt.Errorf("failed to compute tapscript root hash: %w",
+			err)
+	}
+
+	// Ensure that the final tapscript root hash matches the computed root
+	// hash.
+	customRoot := g.customSubtreeRoot.UnwrapOr(chainhash.Hash{})
+
+	if !g.root.IsEqual(&tapscript.root) {
+		return fmt.Errorf("failed to derive tapscript root from "+
+			"internal key, genesis asset ID, and "+
+			"custom subtree root (expected_root=%s, "+
+			"computed_root=%s, custom_subtree_root=%s, "+
+			"genesis_asset_id=%x)",
+			g.root, tapscript.root, customRoot, assetID[:])
+	}
+
+	return nil
+}
+
+// GroupKeyRevealV1 is a version 1 group key reveal type for representing the
+// data used to derive and verify the tweaked key used to identify an asset
+// group.
+type GroupKeyRevealV1 struct {
+	// version is the version of the group key reveal.
+	version uint8
+
+	// internalKey refers to the internal key used to derive the asset
+	// group key. Typically, this internal key is the user's signing public
+	// key.
+	internalKey SerializedKey
+
+	// tapscript is the tapscript tree that commits to the genesis asset ID
+	// and any script spend conditions for the group key.
+	tapscript GroupKeyRevealTapscript
+}
+
+// Ensure that GroupKeyRevealV1 implements the GroupKeyReveal interface.
+var _ GroupKeyReveal = (*GroupKeyRevealV1)(nil)
+
+// NewGroupKeyRevealV1 creates a new version 1 group key reveal instance.
+func NewGroupKeyRevealV1(internalKey btcec.PublicKey,
+	genesisAssetID ID,
+	customRoot fn.Option[chainhash.Hash]) lfn.Result[GroupKeyRevealV1] {
+
+	// Compute the final tapscript root.
+	gkrTapscript, err := NewGroupKeyTapscriptRoot(
+		genesisAssetID, customRoot,
+	).Unpack()
+	if err != nil {
+		return lfn.Err[GroupKeyRevealV1](err)
+	}
+
+	return lfn.Ok(GroupKeyRevealV1{
+		version:     1,
+		internalKey: ToSerialized(&internalKey),
+		tapscript:   gkrTapscript,
+	})
+}
+
+// ScriptSpendControlBlock returns the control block for the script spending
+// path in the custom tapscript subtree.
+//
+// nolint: lll
+func (g *GroupKeyRevealV1) ScriptSpendControlBlock(
+	genesisAssetID ID) lfn.Result[txscript.ControlBlock] {
+
+	internalKey, err := btcec.ParsePubKey(g.internalKey[:])
+	if err != nil {
+		return lfn.Errf[txscript.ControlBlock]("failed to parse "+
+			"internal key: %w", err)
+	}
+
+	outputKey := txscript.ComputeTaprootOutputKey(
+		internalKey, g.tapscript.root[:],
+	)
+	outputKeyIsOdd := outputKey.SerializeCompressed()[0] == 0x03
```

Comment:
nit: use the `secp256k1.PubKeyFormatCompressedOdd` constant (can create a local copy if things get too long otherwise). Both here and in the test.